src/nasuta-sebastian/Fsd.Sebastian.Cs/Fsd.Sebastian.Cs2/AnimalFactory.cs

```diff
@@ -6,47 +6,32 @@ namespace Fsd.Sebastian.Cs2
 {
     public class AnimalFactory
     {
-        private List<Animal> _animals;
-
-        private int numberOfAnimals;
-
-        private string animalType;
+        private List<IAnimalPresenter> _animals;
 
         public void Start()
         {
-            _animals = new List<Animal>();
-
-            animalType = "cat";
-            numberOfAnimals = SetAnimalsCount(animalType);
-            for (int i = 0; i < numberOfAnimals; i++)
-                _animals.Add(new Cat(ConsolePresenter.GetString($"{animalType} name: ")));
-
-            animalType = "dog";
-            numberOfAnimals = SetAnimalsCount(animalType);
-            for (int i = 0; i < numberOfAnimals; i++)
-                _animals.Add(new Dog(ConsolePresenter.GetString($"{animalType} name: ")));
-
-            animalType = "cow";
-            numberOfAnimals = SetAnimalsCount(animalType);
-            for (int i = 0; i < numberOfAnimals; i++)
-                _animals.Add(new Cow(ConsolePresenter.GetString($"{animalType} name: ")));
+            _animals = new List<IAnimalPresenter>();
 
-            animalType = "sheep";
-            numberOfAnimals = SetAnimalsCount(animalType);
-            for (int i = 0; i < numberOfAnimals; i++)
-                _animals.Add(new Sheep(ConsolePresenter.GetString($"{animalType} name: ")));
+            _animals.AddRange(CreateAnimals<Cat>("cat"));
+            _animals.AddRange(CreateAnimals<Dog>("dog"));
+            _animals.AddRange(CreateAnimals<Cow>("cow"));
+            _animals.AddRange(CreateAnimals<Sheep>("sheep"));
 
             foreach (var animal in _animals)
             {
                 ConsolePresenter.PrintString(animal.SayHello());
             }
         }
 
-        private int SetAnimalsCount(string speciesType)
+        private List<IAnimalPresenter> CreateAnimals<T>(string animalType) where T : IAnimalPresenter
         {
-            int numberOfAnimals = int.Parse(ConsolePresenter.GetString($"Insert number of {speciesType}s: "));
+            List<IAnimalPresenter> animals = new List<IAnimalPresenter>();
+            int numberOfAnimals = ConsolePresenter.GetInt($"Insert number of {animalType}s: ");
 
-            return numberOfAnimals;
+            for (var i = 0; i < numberOfAnimals; i++)
+                animals.Add((T)Activator.CreateInstance(typeof(T), ConsolePresenter.GetString($"{animalType} name: ")));
```

Comment:
bardzo spoko. Co prawda troche to wychodziło poza poziom kursu (mogłeś to zrobić switchem) ale wielki plus za użycie Aktywatora.

Reply:
No właśnie nie ogarniam jak to zrobić switchem przy tych zmieniających się typach, tak żeby znowu niepotrzebnie nie kopiować kodu.
W zasadzie już w sobotę, przy wysyłaniu pierwszej wersji wiedziałem, że ten fragment kodu jest do poprawy. Też mi się te powielenia nie podobały, ale póki tego aktywatora nie znalazłem to nie wiedziałem jak to rozwiązać.
Może jak przez weekend będzie wolna chwila to mi to wytłumaczysz.